diff --git a/DB.py b/DB.py
index e25a702e..42a1e60f 100644
--- a/DB.py
+++ b/DB.py
@@ -24,6 +24,9 @@ import hdf5
 import threading
 import multiprocessing
 
+from influxdb import DataFrameClient
+from influxdb import InfluxDBClient
+
 j = 0
 class dbObject:
     def __init__(self, **obj):
@@ -64,6 +67,17 @@ def close_db():
 def close_db_client(c):
     c.close()
 
+def open_influx_client(country):
+    if country == 'US':
+        database='US_Stocks'
+    else:
+        database = 'India_Stocks'
+    client = DataFrameClient('localhost', 8086, 'vpetla', 'petla123', database)
+    return client
+
+def close_influx_client(c):
+    c.close()
+
 def ignore_stock(stk):
     if 'trading' in stk['bscs'].keys():
         if stk['bscs']['trading'] == 'NO' or stk['bscs']['trading'] == 'No':
diff --git a/US_analysis.py b/US_analysis.py
index 40b22022..0ddd3a96 100644
--- a/US_analysis.py
+++ b/US_analysis.py
@@ -25,7 +25,8 @@ def US_main():
     #DB.update_sector_info()
     #internet.price_surprises('US', 0.10, WEEK | DAY, 'COLD', 'SYNC_DB')
     #DB.set_sno('US')
-    hdf5.insert_all_dfs_from_hdf_to_sql('US')
+    #hdf5.insert_all_dfs_from_hdf_to_sql('US')
+    hdf5.insert_all_dfs_from_hdf_to_influxdb('US')
     #internet.send_email("Hello World")
     #internet.send_email2('petlafin@gmail.com', 'Tasche3#Gm', 'petlafin@gmail.com', 'Test', 'Hello')
     #get_radar_stocks()
diff --git a/hdf5.py b/hdf5.py
index bc6d19bc..273cf1ff 100644
--- a/hdf5.py
+++ b/hdf5.py
@@ -195,6 +195,26 @@ def insert_all_dfs_from_hdf_to_sql(country):
             insert_df_from_hdf_to_sql(country, symbol, engine, table)
         i = i + 1
 
+def insert_df_from_hdf_to_influxdb(symbol, df, client):
+    df['Date'] = df.index.strftime("%Y-%m-%d")
+    client.write_points(df, symbol, tag_columns=['Date'], protocol='line', batch_size=1000)
+
+def insert_all_dfs_from_hdf_to_influxdb(country):
+    symbols = get_symbols_from_hdf(country)
+
+    for i, symbol in enumerate(symbols):
+        print("%d: %r"%(i, symbol))
+        df = read_from_hdf(country, symbol)
+        try:
+            client = DB.open_influx_client(country)
+            insert_df_from_hdf_to_influxdb(symbol, df, client)
+        finally:
+            DB.close_influx_client(client)
+        i = i + 1
+        if i > 100:
+            break
+
+
 #def write_to_hdf_store(country, df, symbol):
 #    if country == 'India':
 #        store = library = 'India_Stocks_Prices'

